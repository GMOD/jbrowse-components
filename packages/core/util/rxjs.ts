import { Observable, Observer } from 'rxjs'
import { takeUntil } from 'rxjs/operators'
import { observeAbortSignal } from '.'

/**
 * Wrapper for rxjs Observable.create with improved error handling and
 * aborting support
 * @param func - observer function, could be async
 */
export function ObservableCreate<T>(
  func: (arg: Observer<T>) => void | Promise<void>,
  signal?: AbortSignal,
): Observable<T> {
  return Observable.create((observer: Observer<T>) => {
    try {
      const ret = func(observer)
      // catch async errors
      if (ret && ret.catch) {
        ret.catch((error: Error) => observer.error(error))
      }
    } catch (error) {
      // catch sync errors
      observer.error(error)
    }
  }).pipe(takeUntil(observeAbortSignal(signal)))
}
