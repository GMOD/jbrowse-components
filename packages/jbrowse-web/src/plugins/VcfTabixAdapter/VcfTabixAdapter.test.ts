import { toArray } from 'rxjs/operators'

import Adapter from './VcfTabixAdapter'

test('adapter can fetch variants from volvox.vcf.gz', async () => {
  const adapter = new Adapter({
    vcfGzLocation: {
      localPath: require.resolve('./test_data/volvox.filtered.vcf.gz'),
    },
    index: {
      index: 'TBI',
      location: {
        localPath: require.resolve('./test_data/volvox.filtered.vcf.gz.tbi'),
      },
    },
  })

  const csiAdapter = new Adapter({
    vcfGzLocation: {
      localPath: require.resolve('./test_data/volvox.filtered.vcf.gz'),
    },
    index: {
      index: 'CSI',
      location: {
        localPath: require.resolve('./test_data/volvox.filtered.vcf.gz.csi'),
      },
    },
  })

  const csiFeatures = await csiAdapter.getFeatures({
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })

  const names = await adapter.getRefNames()
  const csiNames = await csiAdapter.getRefNames()
  expect(names).toEqual(csiNames)
  expect(names).toMatchSnapshot()

  const features = await adapter.getFeatures({
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })

  const featuresArray = await features.pipe(toArray()).toPromise()
  const csiFeaturesArray = await csiFeatures.pipe(toArray()).toPromise()
  expect(featuresArray.slice(0, 5)).toMatchSnapshot()
  expect(csiFeaturesArray.slice(0, 5)).toEqual(featuresArray.slice(0, 5))

  const featuresNonExist = await adapter.getFeatures({
    refName: 'ctgC',
    start: 0,
    end: 20000,
  })

  const featuresArrayNonExist = await featuresNonExist
    .pipe(toArray())
    .toPromise()
  expect(featuresArrayNonExist).toEqual([])
})
