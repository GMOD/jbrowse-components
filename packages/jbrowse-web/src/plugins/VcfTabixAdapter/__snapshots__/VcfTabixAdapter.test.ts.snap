// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch variants from volvox.vcf.gz 1`] = `
Array [
  "ctgA",
]
`;

exports[`adapter can fetch variants from volvox.vcf.gz 2`] = `
Array [
  Object {
    "AC1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele count (no HWE assumption)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        2,
      ],
    },
    "AF1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele frequency (assuming HWE)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        1,
      ],
    },
    "DP": Object {
      "meta": Object {
        "Description": "Raw read depth",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        3,
      ],
    },
    "DP4": Object {
      "meta": Object {
        "Description": "# high-quality ref-forward bases, ref-reverse, alt-forward and alt-reverse bases",
        "Number": 4,
        "Type": "Integer",
      },
      "values": Array [
        0,
        0,
        3,
        0,
      ],
    },
    "FQ": Object {
      "meta": Object {
        "Description": "Phred probability of all samples being the same",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        -36,
      ],
    },
    "MQ": Object {
      "meta": Object {
        "Description": "Root-mean-square mapping quality of covering reads",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        37,
      ],
    },
    "VDB": Object {
      "meta": Object {
        "Description": "Variant Distance Bias",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        0.0186,
      ],
    },
    "aliases": undefined,
    "alternative_alleles": Object {
      "meta": Object {
        "description": "VCF ALT field, list of alternate non-reference alleles called on at least one of the samples",
      },
      "values": Array [
        "C",
      ],
    },
    "description": "SNV T -> C",
    "end": 277,
    "filters": Object {
      "meta": Object {
        "description": "List of filters that this site has not passed, or PASS if it has passed all filters",
        "filters": Object {
          "PASS": Object {
            "Description": "Passed all filters",
          },
        },
      },
      "values": null,
    },
    "name": undefined,
    "refName": "ctgA",
    "reference_allele": "T",
    "score": 10.4,
    "start": 276,
    "type": "SNV",
    "uniqueId": "vcf-2560",
  },
  Object {
    "AC1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele count (no HWE assumption)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        2,
      ],
    },
    "AF1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele frequency (assuming HWE)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        1,
      ],
    },
    "DP": Object {
      "meta": Object {
        "Description": "Raw read depth",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        23,
      ],
    },
    "DP4": Object {
      "meta": Object {
        "Description": "# high-quality ref-forward bases, ref-reverse, alt-forward and alt-reverse bases",
        "Number": 4,
        "Type": "Integer",
      },
      "values": Array [
        0,
        1,
        8,
        14,
      ],
    },
    "FQ": Object {
      "meta": Object {
        "Description": "Phred probability of all samples being the same",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        -73,
      ],
    },
    "MQ": Object {
      "meta": Object {
        "Description": "Root-mean-square mapping quality of covering reads",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        37,
      ],
    },
    "PV4": Object {
      "meta": Object {
        "Description": "P-values for strand bias, baseQ bias, mapQ bias and tail distance bias",
        "Number": 4,
        "Type": "Float",
      },
      "values": Array [
        1,
        1,
        0.42,
        1,
      ],
    },
    "VDB": Object {
      "meta": Object {
        "Description": "Variant Distance Bias",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        0.0308,
      ],
    },
    "aliases": undefined,
    "alternative_alleles": Object {
      "meta": Object {
        "description": "VCF ALT field, list of alternate non-reference alleles called on at least one of the samples",
      },
      "values": Array [
        "C",
      ],
    },
    "description": "SNV T -> C",
    "end": 1694,
    "filters": Object {
      "meta": Object {
        "description": "List of filters that this site has not passed, or PASS if it has passed all filters",
        "filters": Object {
          "PASS": Object {
            "Description": "Passed all filters",
          },
        },
      },
      "values": null,
    },
    "name": undefined,
    "refName": "ctgA",
    "reference_allele": "T",
    "score": 107,
    "start": 1693,
    "type": "SNV",
    "uniqueId": "vcf-2658",
  },
  Object {
    "AC1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele count (no HWE assumption)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        1,
      ],
    },
    "AF1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele frequency (assuming HWE)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        0.5,
      ],
    },
    "DP": Object {
      "meta": Object {
        "Description": "Raw read depth",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        22,
      ],
    },
    "DP4": Object {
      "meta": Object {
        "Description": "# high-quality ref-forward bases, ref-reverse, alt-forward and alt-reverse bases",
        "Number": 4,
        "Type": "Integer",
      },
      "values": Array [
        6,
        5,
        3,
        8,
      ],
    },
    "FQ": Object {
      "meta": Object {
        "Description": "Phred probability of all samples being the same",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        45.3,
      ],
    },
    "MQ": Object {
      "meta": Object {
        "Description": "Root-mean-square mapping quality of covering reads",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        37,
      ],
    },
    "PV4": Object {
      "meta": Object {
        "Description": "P-values for strand bias, baseQ bias, mapQ bias and tail distance bias",
        "Number": 4,
        "Type": "Float",
      },
      "values": Array [
        0.39,
        1,
        1,
        1,
      ],
    },
    "VDB": Object {
      "meta": Object {
        "Description": "Variant Distance Bias",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        0.0253,
      ],
    },
    "aliases": undefined,
    "alternative_alleles": Object {
      "meta": Object {
        "description": "VCF ALT field, list of alternate non-reference alleles called on at least one of the samples",
      },
      "values": Array [
        "G",
      ],
    },
    "description": "SNV T -> G",
    "end": 2644,
    "filters": Object {
      "meta": Object {
        "description": "List of filters that this site has not passed, or PASS if it has passed all filters",
        "filters": Object {
          "PASS": Object {
            "Description": "Passed all filters",
          },
        },
      },
      "values": null,
    },
    "name": undefined,
    "refName": "ctgA",
    "reference_allele": "T",
    "score": 44,
    "start": 2643,
    "type": "SNV",
    "uniqueId": "vcf-2775",
  },
  Object {
    "AC1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele count (no HWE assumption)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        2,
      ],
    },
    "AF1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele frequency (assuming HWE)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        1,
      ],
    },
    "DP": Object {
      "meta": Object {
        "Description": "Raw read depth",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        19,
      ],
    },
    "DP4": Object {
      "meta": Object {
        "Description": "# high-quality ref-forward bases, ref-reverse, alt-forward and alt-reverse bases",
        "Number": 4,
        "Type": "Integer",
      },
      "values": Array [
        0,
        0,
        8,
        11,
      ],
    },
    "FQ": Object {
      "meta": Object {
        "Description": "Phred probability of all samples being the same",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        -84,
      ],
    },
    "MQ": Object {
      "meta": Object {
        "Description": "Root-mean-square mapping quality of covering reads",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        36,
      ],
    },
    "VDB": Object {
      "meta": Object {
        "Description": "Variant Distance Bias",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        0.0384,
      ],
    },
    "aliases": undefined,
    "alternative_alleles": Object {
      "meta": Object {
        "description": "VCF ALT field, list of alternate non-reference alleles called on at least one of the samples",
      },
      "values": Array [
        "A",
      ],
    },
    "description": "SNV T -> A",
    "end": 3213,
    "filters": Object {
      "meta": Object {
        "description": "List of filters that this site has not passed, or PASS if it has passed all filters",
        "filters": Object {
          "PASS": Object {
            "Description": "Passed all filters",
          },
        },
      },
      "values": null,
    },
    "name": undefined,
    "refName": "ctgA",
    "reference_allele": "T",
    "score": 124,
    "start": 3212,
    "type": "SNV",
    "uniqueId": "vcf-2892",
  },
  Object {
    "AC1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele count (no HWE assumption)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        2,
      ],
    },
    "AF1": Object {
      "meta": Object {
        "Description": "Max-likelihood estimate of the first ALT allele frequency (assuming HWE)",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        1,
      ],
    },
    "DP": Object {
      "meta": Object {
        "Description": "Raw read depth",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        15,
      ],
    },
    "DP4": Object {
      "meta": Object {
        "Description": "# high-quality ref-forward bases, ref-reverse, alt-forward and alt-reverse bases",
        "Number": 4,
        "Type": "Integer",
      },
      "values": Array [
        0,
        0,
        5,
        6,
      ],
    },
    "FQ": Object {
      "meta": Object {
        "Description": "Phred probability of all samples being the same",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        -67.5,
      ],
    },
    "INDEL": Object {
      "meta": Object {
        "Description": "Indicates that the variant is an INDEL.",
        "Number": 0,
        "Type": "Flag",
      },
      "values": null,
    },
    "MQ": Object {
      "meta": Object {
        "Description": "Root-mean-square mapping quality of covering reads",
        "Number": 1,
        "Type": "Integer",
      },
      "values": Array [
        37,
      ],
    },
    "VDB": Object {
      "meta": Object {
        "Description": "Variant Distance Bias",
        "Number": 1,
        "Type": "Float",
      },
      "values": Array [
        0.0384,
      ],
    },
    "aliases": undefined,
    "alternative_alleles": Object {
      "meta": Object {
        "description": "VCF ALT field, list of alternate non-reference alleles called on at least one of the samples",
      },
      "values": Array [
        "ct",
      ],
    },
    "description": "deletion ctt -> ct",
    "end": 3860,
    "filters": Object {
      "meta": Object {
        "description": "List of filters that this site has not passed, or PASS if it has passed all filters",
        "filters": Object {
          "PASS": Object {
            "Description": "Passed all filters",
          },
        },
      },
      "values": null,
    },
    "name": undefined,
    "refName": "ctgA",
    "reference_allele": "ctt",
    "score": 160,
    "start": 3857,
    "type": "deletion",
    "uniqueId": "vcf-2994",
  },
]
`;
